File: src/backend/XpertSphere.MonolithApi/Models/Application.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using System.Text.Json.Serialization;
using XpertSphere.MonolithApi.Enums;
using XpertSphere.MonolithApi.Models.Base;

namespace XpertSphere.MonolithApi.Models;

public class Application : AuditableEntity
{
    [MaxLength(2000)]
    public string? CoverLetter { get; set; }

    [MaxLength(1000)]
    public string? AdditionalNotes { get; set; }

    [Required]
    public ApplicationStatus CurrentStatus { get; set; } = ApplicationStatus.Applied;

    [Range(1, 5)]
    public int? Rating { get; set; }

    [Required]
    public DateTime AppliedAt { get; set; } = DateTime.UtcNow;

    public DateTime? LastUpdatedAt { get; set; }

    [Required]
    public Guid JobOfferId { get; set; }

    [Required]
    public Guid CandidateId { get; set; }

    public Guid? AssignedTechnicalEvaluatorId { get; set; }

    public Guid? AssignedManagerId { get; set; }

    // Navigation properties
    [ForeignKey("JobOfferId")]
    [JsonIgnore]
    public virtual JobOffer JobOffer { get; set; } = null!;

    [ForeignKey("CandidateId")]
    [JsonIgnore]
    public virtual User Candidate { get; set; } = null!;

    [ForeignKey("AssignedTechnicalEvaluatorId")]
    [JsonIgnore]
    public virtual User? AssignedTechnicalEvaluator { get; set; }

    [ForeignKey("AssignedManagerId")]
    [JsonIgnore]
    public virtual User? AssignedManager { get; set; }

    [JsonIgnore]
    public virtual ICollection<ApplicationStatusHistory> StatusHistory { get; set; } = [];

    // Computed properties
    [NotMapped]
    public bool IsActive => CurrentStatus != ApplicationStatus.Rejected && 
                           CurrentStatus != ApplicationStatus.Withdrawn && 
                           CurrentStatus != ApplicationStatus.Accepted;

    [NotMapped]
    public bool IsCompleted => CurrentStatus == ApplicationStatus.Accepted || 
                              CurrentStatus == ApplicationStatus.Rejected || 
                              CurrentStatus == ApplicationStatus.Withdrawn;

    [NotMapped]
    public bool IsInProgress => CurrentStatus != ApplicationStatus.Applied && IsActive;

    [NotMapped]
    public string StatusDisplayName => CurrentStatus switch
    {
        ApplicationStatus.Applied => "Application Submitted",
        ApplicationStatus.Reviewed => "Under Review",
        ApplicationStatus.PhoneScreening => "Phone Screening",
        ApplicationStatus.TechnicalTest => "Technical Assessment",
        ApplicationStatus.TechnicalInterview => "Technical Interview",
        ApplicationStatus.FinalInterview => "Final Interview",
        ApplicationStatus.OfferMade => "Offer Extended",
        ApplicationStatus.Accepted => "Offer Accepted",
        ApplicationStatus.Rejected => "Application Rejected",
        ApplicationStatus.Withdrawn => "Application Withdrawn",
        _ => CurrentStatus.ToString()
    };

}

File: src/frontend/packages/candidate-app/src/models/application.ts
import { ApplicationStatus, ApplicationSource } from '../enums';

export interface Application {
  id: string;
  candidateId: string;
  candidateName: string;
  candidateEmail: string;
  jobId: string;
  jobTitle: string;
  status: ApplicationStatus;
  appliedAt: string;
  updatedAt: string;
  source: ApplicationSource;
  coverLetter: string;
  resumeUrl: string;
  rating: number | null;
  notes: string;
}

export interface ApplicationStatusConfig {
  color: string;
  textColor: string;
  icon: string;
  label: string;
}

export const applicationStatusConfig: Record<ApplicationStatus, ApplicationStatusConfig> = {
  [ApplicationStatus.Applied]: {
    color: 'grey',
    textColor: 'white',
    icon: 'send',
    label: 'Candidature déposée',
  },
  [ApplicationStatus.Reviewed]: {
    color: 'blue',
    textColor: 'white',
    icon: 'rate_review',
    label: 'Examinée',
  },
  [ApplicationStatus.PhoneScreening]: {
    color: 'cyan',
    textColor: 'white',
    icon: 'phone',
    label: 'Entretien téléphonique',
  },
  [ApplicationStatus.TechnicalTest]: {
    color: 'indigo',
    textColor: 'white',
    icon: 'code',
    label: 'Test technique',
  },
  [ApplicationStatus.TechnicalInterview]: {
    color: 'purple',
    textColor: 'white',
    icon: 'developer_mode',
    label: 'Entretien technique',
  },
  [ApplicationStatus.FinalInterview]: {
    color: 'deep-purple',
    textColor: 'white',
    icon: 'groups',
    label: 'Entretien final',
  },
  [ApplicationStatus.OfferMade]: {
    color: 'teal',
    textColor: 'white',
    icon: 'local_offer',
    label: 'Offre faite',
  },
  [ApplicationStatus.Accepted]: {
    color: 'positive',
    textColor: 'white',
    icon: 'check_circle',
    label: 'Acceptée',
  },
  [ApplicationStatus.Rejected]: {
    color: 'negative',
    textColor: 'white',
    icon: 'cancel',
    label: 'Rejetée',
  },
  [ApplicationStatus.Withdrawn]: {
    color: 'warning',
    textColor: 'white',
    icon: 'remove_circle',
    label: 'Retirée',
  },
};

export const applicationSourceLabels: Record<ApplicationSource, string> = {
  [ApplicationSource.Website]: 'Site web',
  [ApplicationSource.LinkedIn]: 'LinkedIn',
  [ApplicationSource.Email]: 'Email',
  [ApplicationSource.Referral]: 'Recommandation',
  [ApplicationSource.Direct]: 'Direct',
};

export interface ApplicationDto {
  id: string;
  coverLetter?: string;
  additionalNotes?: string;
  currentStatus: ApplicationStatus;
  statusDisplayName: string;
  rating?: number;
  appliedAt: string;
  lastUpdatedAt?: string;
  createdAt: string;
  updatedAt?: string;

  jobOfferId: string;
  jobOfferTitle: string;
  organizationName: string;

  candidateId: string;
  candidateName: string;
  candidateEmail: string;

  assignedTechnicalEvaluatorId?: string;
  assignedTechnicalEvaluatorName?: string;

  assignedManagerId?: string;
  assignedManagerName?: string;

  isActive: boolean;
  isCompleted: boolean;
  isInProgress: boolean;

  statusHistory: ApplicationStatusHistoryDto[];
}

export interface CreateApplicationDto {
  jobOfferId: string;
  coverLetter?: string;
  additionalNotes?: string;
}

export interface UpdateApplicationDto {
  coverLetter?: string;
  additionalNotes?: string;
  rating?: number;
}

export interface ApplicationStatusHistoryDto {
  id: string;
  applicationId: string;
  status: ApplicationStatus;
  statusDisplayName: string;
  comment: string;
  rating?: number;
  ratingDescription: string;
  hasRating: boolean;
  updatedByUserId: string;
  updatedByUserName: string;
  updatedAt: string;
}

export interface WithdrawApplicationRequest {
  reason: string;
}
